### PR TITLE
Historical query perf improvements: Replace signature verification with GCM header check

### PR DESCRIPTION
**Pull request by eddyashton**
Signature verification is only required to confirm provenance. To confirm integrity, it is sufficient to validate the existing GCM header. Providence is not a meaningful property for nodes to confirm, instead they can return receipts which allow callers to check signatures + provenance if they wish.

An additional check which we didn't have to do previously but do now is explicitly confirming that each entry is a precursor to the current state, by checking its (integrity-protected) TxID matches the consensus' view history.

TODO:
- [x] The flow inside `historical_queries.h` is now over-complicated, since it only trusts each store once it has seen the following signature. This should be fixed.
- [x] We can go even faster by avoiding the signature fetch when the receipt is not requested. This requires some API change to indicate a request where the receipt is optional.

### CODE REVIEW

**Comment by achamayou** [comment]
@eddyashton both TODOs are useful, but are best done in separate PRs, because I suspect they will yield little benefit in the pathologically signature-dense cases we're primarily looking to improve at the moment.

**Comment by eddyashton** [comment]
> @eddyashton both TODOs are useful, but are best done in separate PRs, because I suspect they will yield little benefit in the pathologically signature-dense cases we're primarily looking to improve at the moment.

@achamayou I've done both of these TODOs already. The API change was actually easy; we already have a distinction at the calling point of whether the caller was looking for a receipt or not - did they call `get_state` (includes receipt) or `get_store` (doesn't). So we just track that through the request's lifetime, and if we're not looking for receipts we don't continue with the signature fetching step.

**Comment by jumaffre** [comment]
image
It looks like the new metric needs a `^` somewhere ;)

File: src/node/historical_queries.h [hunk 1]
```diff
@@ -386,47 +342,6 @@ namespace ccf::historical
       }
     }
 
-    // Returns true if this is a valid signature that passes our verification
-    // checks
-    bool verify_signature(const StorePtr& sig_store, ccf::SeqNo sig_seqno)
```
Comment (jumaffre) [inline src/node/historical_queries.h:391]: Correct me if I'm wrong but I think we can actually delete retired nodes from the store now, as historical queries were the only reason we couldn't complete this work before (see #2474).

Comment (achamayou) [reply src/node/historical_queries.h:391]: That's correct, but best done in a separate PR.

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
I think it'd be really good to add an end to end test along the following lines:

1. Start a three nodes network.
2. Isolate the primary.
3. Run transactions against both the old primary and the new, enough to produce a chunk on each side.
4. Copy the chunk from the old primary to one of the nodes in the partition
5. Attempt to run a historical query on that node in the range covered by the chunk, and confirm that it fails

**Pull request merged**